### PR TITLE
tests: Rework tests and enable integration tests

### PR DESCRIPTION
**Pull request by aarnq**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

This adds test helpers to setup local clusters in integration tests, and tests for local clusters itself and converted harbor tests as a poc.

To facilitate this the harness has been fully containerised through the use of the `run-from-container.sh` script, which works with both docker and podman. 
Tested to work with `cypress open` and it can open the UI.
Tested to work with different UIDs on the host then those of the container (1000).
GitHub Actions uses a different one, so as it works there it works with anything.

Now changes to tests does not require changes to the tests image:
- The build process has changed so that no artefacts from the repository is required during build.
- The generation process has changed so that it is not within the build process.
- The dependencies has become globals in the container image, instead of being installed by make.

The library structure for bats tests have changed and the main library file now has helpers to load in additional libraries, including external ones.

Changes to `local-cluster.sh` to provide local resolve for local clusters. 
It spins up a CoreDNS container to provide static entries and upstream DNS in a more managed way then relying on `node-local-dns`, this was required to ensure tests work as intended in GitHub Actions.

#### Information to reviewers

To test unit and integration tests:

```
# From tests/

# First start local cache else you will be rate limited.
../scripts/local-cluster.sh cache create
# Second start local resolve (will prompt for password to set DNS if you have resolvectl).
../scripts/local-cluster.sh resolve create integration.dev-ck8s.com

# Build container
make build-unit # for unit tests
make build-main # for other tests

# Takes quite a while
make run-unit
make run-integration
# To run only api (bats) tests on harbor
make run-integration/harbor/use-api.bats
# To run only ui (cypress) tests on harbor
make run-integration/harbor/use-ui.bats
```

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests
- Config checks:
  - [ ] The schema was updated

### CODE REVIEW

**Comment by aarnq** [comment]
[The first successful run of integration tests](https://github.com/elastisys/compliantkubernetes-apps/actions/runs/9130811573/job/25108581169) from [the containerised test harness](https://github.com/elastisys/compliantkubernetes-apps/pull/2120/commits/1c983a2f1bd6a97a867f8ebe8ef7671a031e2be1) :tada:
_And it broke unit tests_ :sweat_smile:

**Comment by aarnq** [comment]
So this now runs unit and integration tests in GitHub Actions so I feel like it is ready for review. Please do test to run this locally, you are required to have rootful docker or rootless podman v5 with the socket enabled, and kind to setup local clusters.

I will setup regression tests to run as integration tests do as well.

I have not tested the current state of the end-to-end tests, but I would like to try to setup minimal deployments in the integration tests to run most of them in a reproducible environment in this PR before merging.

**Comment by lunkan93** [comment]
Managed to run the integration tests locally:

```console
❯ make run-integration
FORWARD_RUNTIME=true ../scripts/run-from-container.sh compliantkubernetes-apps-tests:main make -C tests internal-integration
[ck8s] forward your environment and/or runtime to container compliantkubernetes-apps-tests:main? [y/N]: y
make: Entering directory '/home/simon/devcluster/compliantkubernetes-apps/tests'
--- bats integration
local-clusters.bats
 ✓ local cluster has kubeconfigs
 ✓ local cluster has ready nodes
 ✓ local cluster has running calico system
 ✓ local cluster has installed minio
 ✓ local cluster has running minio
 ✓ local cluster has installed tigera operator
 ✓ local cluster has running tigera operator

7 tests, 0 failures

make: Leaving directory '/home/simon/devcluster/compliantkubernetes-apps/tests'
```

Only really had to give r/w permissions for all users on the docker socket to get it working.

```bash
sudo chmod 666 /var/run/docker.sock
```

Nice work :tada:

**Comment by aarnq** [comment]
Ready for final review.

**Review by Zash — APPROVED** [review]

File: tests/README.md [hunk 1]
```diff
@@ -1,52 +1,54 @@
 # Apps Tests
 
 The test suite is implemented using [`bats`](https://github.com/bats-core/bats-core) and [`cypress`](https://github.com/cypress-io/cypress), with unit, regression, integration, and end-to-end tests under their own respective directory.
-Tests implemented with `cypress` can and will be integrated into the `bats` test suite using generators.
+
+Generators are employed to generate `bats` tests from `cypress` and `gotmpl` tests to integrate it into the rest of test suites.
```
Comment (simonklb) [inline tests/README.md:5]: This was a bit hard to parse. What about something like this?
```suggestion
For each `cypress` test a `bats` test is generated that runs the `cypress` test and verifies the result.
Some tests are also generated from `gotmpl` templates.
```

Comment (simonklb) [reply tests/README.md:5]: On the other hand there is already good descriptions of the different tests that you can write and the code generation further down so maybe this can be removed?

File: tests/README.md [hunk 2]
```diff
@@ -1,52 +1,54 @@
...
+> Do not use `docker` or `podman` directly from tests, as they take differing arguments and may depend on variables present on the host not available within the tests container.
+> Instead use tools like `buildah` for building and pushing or `skopeo` for syncing and pulling.
 >
 > Additionally tests running in the test container might hang on interrupts, requiring the container to be killed.
 
 The tests differentiate between static and dynamic tests, all static tests can be run without setting up an environment, and all dynamic tests requires an environment to test.
 Static tests are tagged with `static`.
 
+> [!danger]
+> This type of distinction will be phased out and all unit, regression, and integration tests will be required to be static!
```
Comment (simonklb) [inline tests/README.md:24]: This doesn't seem to render when I preview the rendered markdown

File: tests/README.md [hunk 3]
```diff
@@ -68,26 +70,24 @@ make run-<unit|regression|integration|end-to-end>
...
 ```bash
 # remove dependencies and generated files
 make clean
 
-# remove dependencies
-make clean-dep
-
 # remove generated files
 make clean-gen
 ```
```
Comment (simonklb) [inline tests/README.md:81]: Can this entire section not be removed now when we commit the generated tests? Or is there still a good reason to clean things up sometimes?

File: tests/README.md [hunk 4]
```diff
@@ -102,27 +102,29 @@ bats <path/to/file.bats>
 
 Additionally tests can be filtered via tags using the `--filter-tags <tags,...>` argument.
```
Comment (simonklb) [inline tests/README.md:103]: This looks odd when it's above the `cypress` instructions. Maybe add a sentence similar to the one above the `bats` instructions or move this?

**Review by simonklb — CHANGES_REQUESTED** [review]
Great work!

Only minor stuff, the only blocker is the link detection since I'm not sure what would happen if it detects more than one.

**Comment by aarnq** [comment]
@viktor-f do you have the time to review this?

**Comment by viktor-f** [comment]
> @viktor-f do you have the time to review this?

Preferably not today, but I can make time for it tomorrow.

**Comment by aarnq** [comment]
Now this is ready for another round of reviews.

**Review by viktor-f — APPROVED** [review]
Tests works well for me to run locally now.
I have not looked thoroughly at all of these changes. Let me know if you want me to check something in particular.

**Review by simonklb — APPROVED** [review]
Fantastic work! :)

File: tests/Makefile [hunk 5]
```diff
@@ -1,229 +1,104 @@
 #!/usr/bin/env make
```
[the remark this answers is not in the document]
Comment (aarnq) [reply tests/Makefile:1]: Yeah, the only thing that is nice about makefile in this instance is the dependency tracking for generated test files, but the general setup is a pain.

Maybe I should add a comment that says _if more complexity is added, write a script instead_. :sweat_smile:

[the remark this answers is not in the document]
Comment (Zash) [reply tests/Makefile:1]: There does exist many [other](https://github.com/casey/just) command runners with dependency tracking, Make is more geared for _making_ things after all :)

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -81,28 +81,12 @@ helmfile -e workload_cluster <operation> --selector app=<application> --include-
 Use `helmfile -e <service|workload>_cluster list` to list all releases and to view their labels.
 By default all releases have `name=<release-name>` and `chart=<chart-name>` as predefined labels.
 
-Enabling ingress and resolve requires a special setup for `ingress-nginx` and `node-local-dns` respectively.
-Both will by default be port-mapped on the local address `127.0.64.43` and once both are running you need to change your computer's network settings to use this address as the DNS resolver, then you can access service endpoints using the configured domain.
+Enabling ingress and resolve requires a special setup.
+The ingress will by default be port-mapped on the local address `127.0.64.43` and the local clusters script provide commands to create and delete local resolve to resolve any domain on and to the same local address.
```
[the remark this answers is not in the document]
Comment (Zash) [reply DEVELOPMENT.md:85]: Maybe "Local DNS resolver", as "DNS server" could be ambiguous when there's authoritative DNS servers and (stub- and recursive-) DNS resolver servers.

[the remark this answers is not in the document]
Comment (aarnq) [reply DEVELOPMENT.md]: Reworded :+1:

File: DEVELOPMENT.md [hunk 7]
```diff
@@ -81,28 +81,12 @@ helmfile -e workload_cluster <operation> --selector app=<application> --include-
...
+Commands to do so is `./scripts/local-cluster.sh resolve <create|delete> <domain>`, matching the base domain of the cluster.
```
[the remark this answers is not in the document]
Comment (aarnq) [reply DEVELOPMENT.md:87]: Done.

File: tests/README.md [hunk 8]
```diff
@@ -1,94 +1,81 @@
...
-make run-<unit|regression|integration|end-to-end>
+make run-<target>                # list targets with make list
+make run-<target>/<suite>        # list targets with make list-<target>
+make run-<target>/<suite>/<file> # list targets with make list-<target>/<suite>
 ```
 
-Additionally tests can be filtered via tags using the `-<tags,...>` suffix to the `run-<target>` command.
-
-Run individual tests:
+To force cypress and template tests to be regenerated run:
```
[the remark this answers is not in the document]
Comment (aarnq) [reply tests/README.md]: Reworded with an if...

File: DEVELOPMENT.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonklb) [reply DEVELOPMENT.md]: The "on and to the same local address" is still a bit hard for me to grasp, but feel free to ignore, it's not a big deal. :smile:

File: tests/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonklb) [reply tests/README.md]: It reads to me that tests are automatically generated, is that really the case? For example, if I add a new test do I not have to also generate the bats wrapper for it myself?

[the remark this answers is not in the document]
Comment (aarnq) [reply tests/README.md]: Yes, if you add a cypress or template then the bats files are generated from them automatically if a source file is newer than the destination file.

File: DEVELOPMENT.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (aarnq) [reply DEVELOPMENT.md]: I missed that part, but now that should be clearer.

File: tests/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonklb) [reply tests/README.md]: When? Clearly not as soon as I save the file? I assume I have to do _something_ to generate the new test files before I commit my changes to the git repository?

[the remark this answers is not in the document]
Comment (aarnq) [reply tests/README.md]: As a regular make prerequisite.

[the remark this answers is not in the document]
Comment (simonklb) [reply tests/README.md]: Ah, that's prefect!
